src/transformStateWithClones.js

```diff
@@ -7,7 +7,39 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateCopy = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    const { type, extraData, keysToRemove } = action;
+
+    switch (type) {
+      case 'addProperties':
+        Object.assign(stateCopy, extraData);
+        break;
+
+      case 'removeProperties':
+        keysToRemove.forEach(function (key) {
+          delete stateCopy[key];
+        });
+        break;
+
+      case 'clear':
+        for (const key in stateCopy) {
+          if (stateCopy.hasOwnProperty(key)) {
+            delete stateCopy[key];
+          }
+        }
+        break;
+
+      default:
+        break;
+    }
+
+    stateHistory.push({ ...stateCopy });
```

Comment:
[DONT REPEAT YOURSELF] - You are correctly pushing the state copy to the history array at the end of each loop iteration, which is good as it avoids repeating this action in each case.